base/libgit2/types.jl

```diff
@@ -0,0 +1,501 @@
+# This file is a part of Julia. License is MIT: http://julialang.org/license
+
+const Cstring_NULL = convert(Cstring, Ptr{UInt8}(C_NULL))
+
+const OID_RAWSZ = 20
+const OID_HEXSZ = OID_RAWSZ * 2
+const OID_MINPREFIXLEN = 4
+
+immutable Oid
+    val::NTuple{OID_RAWSZ, UInt8}
+    Oid(val::NTuple{OID_RAWSZ, UInt8}) = new(val)
+end
+Oid() = Oid(ntuple(i->zero(UInt8), OID_RAWSZ))
+
+immutable TimeStruct
+    time::Int64     # time in seconds from epoch
+    offset::Cint    # timezone offset in minutes
+end
+TimeStruct() = TimeStruct(zero(Int64), zero(Cint))
+
+immutable SignatureStruct
+    name::Ptr{UInt8}  # full name of the author
+    email::Ptr{UInt8} # email of the author
+    when::TimeStruct  # time when the action happened
+end
+SignatureStruct() = SignatureStruct(Ptr{UInt8}(0),
+                                    Ptr{UInt8}(0),
+                                    TimeStruct())
+
+immutable StrArrayStruct
+   strings::Ptr{Cstring}
+   count::Csize_t
+end
+StrArrayStruct() = StrArrayStruct(Ptr{Cstring}(C_NULL), zero(Csize_t))
+function Base.finalize(sa::StrArrayStruct)
+    sa_ptr = Ref(sa)
+    ccall((:git_strarray_free, :libgit2), Void, (Ptr{StrArrayStruct},), sa_ptr)
+    return sa_ptr[]
+end
+
+immutable Buffer
+    ptr::Ptr{Cchar}
+    asize::Csize_t
+    size::Csize_t
+end
+Buffer() = Buffer(Ptr{Cchar}(C_NULL), zero(Csize_t), zero(Csize_t))
+function Base.finalize(buf::Buffer)
+    buf_ptr = Ref(buf)
+    ccall((:git_buf_free, :libgit2), Void, (Ptr{Buffer},), buf_ptr)
+    return buf_ptr[]
+end
+
+immutable CheckoutOptions
+    version::Cuint
+
+    checkout_strategy::Cuint
+
+    disable_filters::Cint
+    dir_mode::Cuint
+    file_mode::Cuint
+    file_open_flags::Cint
+
+    notify_flags::Cuint
+    notify_cb::Ptr{Void}
+    notify_payload::Ptr{Void}
+
+    progress_cb::Ptr{Void}
+    progress_payload::Ptr{Void}
+
+    paths::StrArrayStruct
+
+    baseline::Ptr{Void}
+    baseline_index::Ptr{Void}
+
+    target_directory::Cstring
+    ancestor_label::Cstring
+    our_label::Cstring
+    their_label::Cstring
+
+    perfdata_cb::Ptr{Void}
+    perfdata_payload::Ptr{Void}
+end
+CheckoutOptions(; checkout_strategy::Cuint = GitConst.CHECKOUT_SAFE,
```

Comment:
All these default-kw immutables could still use a little macro I think to cut down on the repetition